docs/tutorials/gluon/learning_rate_finder.md

```diff
@@ -0,0 +1,322 @@
+                    batch = []
+
+sampler = mx.gluon.data.RandomSampler(len(dataset))
+batch_sampler = ContinuousBatchSampler(sampler, batch_size=128)
+data_loader = mx.gluon.data.DataLoader(dataset, batch_sampler=batch_sampler, num_workers=cpu_count())
+```
+
+## Implementation
+
+With preparation complete, we're ready to write our Learning Rate Finder that wraps the `Learner` we defined above. We implement a `find` method for the procedure, and `plot` for the visualization. Starting with a very low learning rate as defined by `lr_start` we train one iteration at a time and keep multiplying the learning rate by `lr_multiplier`. We analyse the loss and continue until it diverges according to `LRFinderStoppingCriteria` (which is defined later on). You may also notice that we save the parameters and state of the optimizer before the process and restore afterwards. This is so the Learning Rate Finder process doesn't impact the state of the model, and can be used at any point during training.
+
+
+```python
+from matplotlib import pyplot as plt
+
+class LRFinder():
+    def __init__(self, learner):
+        """
+        learner: able to take single iteration with given learning rate and return loss
+           and save and load parameters of the network (Learner)
+        """
+        self.learner = learner
+        
+    def find(self, lr_start=1e-6, lr_multiplier=1.1, smoothing=0.3):
+        """
+        lr_start: learning rate to start search (float)
+        lr_multiplier: factor the learning rate is multiplied by at each step of search (float)
+        smoothing: amount of smoothing applied to loss for stopping criteria (float)
+        """
+        # Used to initialize weights; pass data, but don't take step.
+        # Would expect for new model with lazy weight initialization
+        self.learner.iteration(take_step=False)
+        # Used to initialize trainer (if no step has been taken)
+        if not self.learner.trainer._kv_initialized:
+            self.learner.trainer._init_kvstore()
+        # Store params and optimizer state for restore after lr_finder procedure
+        # Useful for applying the method partway through training, not just for initialization of lr.
+        self.learner.net.save_params("lr_finder.params")
+        self.learner.trainer.save_states("lr_finder.state")
+        lr = lr_start
+        self.results = [] # List of (lr, loss) tuples
+        stopping_criteria = LRFinderStoppingCriteria(smoothing)
+        while True:
+            # Run iteration, and block until loss is calculated.
+            loss = self.learner.iteration(lr).asscalar()
+            self.results.append((lr, loss))
+            if stopping_criteria(loss):
+                break
+            lr = lr * lr_multiplier
+        # Restore params (as finder changed them)
+        self.learner.net.load_params("lr_finder.params", ctx=self.learner.ctx)
+        self.learner.trainer.load_states("lr_finder.state")
+        self.plot()
```

Comment:
can you return results here rather than calling plot? Users might not be a in a graphical environment

Reply:
Removed plot, returned results. Called plot separately.